
syntax = "proto3";

package starflow.proto;

service UIService {
    rpc SayHello (HelloRequest) returns (HelloReply) { }
}

message HelloRequest { }

message HelloReply { }

message packet {
    int64 ts_in       = 1;
    int64 ts_out      = 2;
    int32 len         = 3;
    int32 qid         = 4;
    int32 qlen        = 5;
    features features = 6;
}

message key {
    int32 ip_proto = 1;
    int32 ip_src   = 2;
    int32 ip_dst   = 3;
    int32 th_sport = 4;
    int32 th_dport = 5;
}

message tcp_flags {
    int32 flags = 1;
}

message features {
    int32 ip_ttl        = 1;
    int32 tcp_seq       = 2;
    tcp_flags tcp_flags = 3;
}

message clfr {
    key key                 = 1;
    bool complete           = 2;
    int64 id                = 3;
    int32 evict_ts_s        = 4;
    int32 table_id          = 5;
    repeated packet packets = 6;
}

message flow_list {
    repeated clfr flows = 1;
}
